Make Security/Command/AI Upload channels have roundstart random frequencies, and Security/Command require an encryption key.

ShiftyRail opened the pull request:
The idea is that people shouldn't be able to metagame those frequencies to spy on security roundstart. If you want to hear what the redshirts say, steal one of their headsets. Or PDA the AI, because it knows them. 

Testing done : 
- [x] Captain's headset, each and every freq of the station
- [x] ERT & Deathsquad
- [x] AIs
- [x] Borgos
- [x]  Pox raiders/Nuclear Operatives/ERT

Done : 
- [x] Randomised frequencies + needing a secure encryption key for the channels. With just the frequency, you can talk, but you cannot hear. Useful for reporting your position with, say, a SBR hidden in your backpack tuned to sec without your abductors noticing shit. 
- [x] Mapping. DJ sat, listening post, and Librarians need to have their encryption keys.

To do : 
- [x] Add some keys somewhere in telecoms + the possibility to reverse-engineer a key.

Non-changelog chances : 
- Reworked Syndicate & Raider channels to less snowflake channels and using proper encryption keys.
- Changed a lot of vareddited broadcasters in maps to proper subtypes of broadcasters.

:cl:
- tweak: Security, AI Upload, the Emergency Response Team, and Command channels have now randomised frequencies instead of fixed values.
- tweak: Security and Command frequencies will not be able to be listened unless you get the proper encryption key.
- tweak: The Librarian, the AI, the HoP and the Captain know all the station's frequencies. Heads of staff know the command frequency. Security know their own frequency, and the IAA/Lawyer know both the security and command frequencies.  
- rsadd: You can now put an encryption key in an intercom to make it decode a frequency range. Be careful, as you'll still need to tune it to the frequency.
- tweak: You can find extra encryption keys in telecomms.

MadmanMartian commented on code/game/communications.dm:
```diff
@@ -153,30 +170,68 @@ var/list/radiochannelsreverse = list(
+	for (var/channel in radiochannelsreverse)
+		to_chat(world, "[channel] = [radiochannelsreverse[channel]]")
+		
+	to_chat(world, "freqtoname :")
```
Remove hitlers when you're done.

ShiftyRail commented on code/game/machinery/telecomms/presets.dm:
```diff
@@ -51,22 +51,23 @@
-			freq_listening |= i
-		..()
+	
+/obj/machinery/telecomms/receiver/preset_right/initialize()
```
Moved the `new()` code into `initialize()` because I didn't feel there was a particular reason for it to be on `new()`

DamianX replied:
But you're not calling initialize in New if the game already started

ShiftyRail replied:
So you mean this should all be in new() ?

DamianX replied:
I'm saying initialize won't run if you create one of these after round start
if for some reason one was to do that

ShiftyRail replied:
I get it. Since it's the `preset` receiver I'm not sure if people will actually create them.

strawberrytan commented:
Will the librarian channel listening guide in his back room still list the frequency or not?

ShiftyRail commented:
@strawberrytan Probably not, but the Librarian will have it in his mind.

DamianX commented:
I don't like this at all. As I've said before, I think encryption keys should be given a purpose instead.

abernd commented:
Make sure you change both station bounced radios and station intercoms to have a manual frequency input (similar to wall mounted sound systems). It would be a pain to switch from 1459 to 1201 on many intercoms just to listen in on people as an AI. 

Keep in mind that all it would take for an assistant to obtain a secure frequency at roundstart would be to PDA any AI or borg.

DamianX commented on code/game/machinery/telecomms/broadcaster.dm:
```diff
@@ -322,33 +322,33 @@ var/message_delay = 0 // To make sure restarting the recentmessages list is kept
-					blackbox.msg_cargo += blackbox_msg
-				else
-					blackbox.messages += blackbox_msg
+			// Sadly can't do a switch() because those expressions aren't constant.
```
use `else if`s then

MadmanMartian replied:
else ifs used.

ShiftyRail commented:
@abernd AIs can choose which transmission mode they use (radios or headset), and they effectively start with the equivalent of a captain's headset. They don't have to skim through the frequencies to talk.

@DamianX I agree that it would be a neater idea. I'll see if I can make a combination of both : you can emit to sec radio if you have the frequency, but can only decode with the encryption key. It'd also solve the issue of asking the AIs the frequencies.

abernd commented:
@ShiftyRail If an AI wanted to change the frequency of a wall mounted station intercom, it would have to manually skim through frequencies from the default (common) to set the desired frequency. Currently the AI private frequency (1447) is very close to common (1459), making it easy to set an intercom to transmit on AI private. It's useful to be able to do this quickly since you can listen in on what people are saying off comms across the station by changing a large number of station intercoms to microphone only on the AI private frequency.

SonixApache commented:
[requires server vote]  probably

abernd commented:
@ShiftyRail That'll fix it. It should be simple enough to change the SBR/intercom UI to allow direct input of a specific frequency, though.

ShiftyRail commented:
I'm going to implement Damian/Sonix's suggestion and make encryption keys play a role.
Obtaining the frequency will just let you shitpost on security without even hearing it.

DamianX commented:
Why?

DrCelt commented on code/modules/paperwork/paper.dm:
```diff
@@ -471,7 +471,7 @@ var/global/list/paper_folding_results = list ( \
 /obj/item/weapon/paper/intercoms
 	name = "paper - 'Ace Reporter Intercom manual'"
-	info = "<B>Welcome new owner!</B><BR><BR>You have purchased the latest in listening equipment. The telecommunication setup we created is the best in listening to common and private radio fequencies.Now to listen to the private channels, you'll have to configure the intercoms.<br> Here is a list of frequencies for you to listen on.<br><ul><li>145.9 - Common Channel</li><li>144.7 - Private AI Chan ...
+	info = "<B>Welcome new owner!</B><BR><BR>You have purchased the latest in listening equipment. The telecommunication setup we created is the best in listening to common and private radio fequencies.Now to listen to the private channels, you'll have to configure the intercoms.<br> The list of frequencies has been passed down orally between Ace Reporters, and you should have it remembered by now."
```
This should have been documented in your PR.

Inorien replied:
>Librarian gives all the greyshits the sec frequency
Finally, Librarian has a use

ShiftyRail replied:
@DrCelt It is now

DrCelt commented:
Also
>With just the frequency, you can talk, but you cannot hear. Useful for reporting your position with, say, a SBR hidden in your backpack tuned to sec without your abductors noticing shit.

As if we need more buffs to security

ShiftyRail commented:
@DrCelt this isn't finished yet, but I agree, we do need it in the changelog.
While it's a security "buff", it's also very simply to make it so you can spy on sec without a smelly metagaming method (looking at the wiki for frequencies). Either steal a headset, PDA the AI for it or fiddle with the frequencies.

Doing it via the encryption keys would be a much stronger from of encryption since it would remove the possibility to steal it via the AI. Which is why I didn't want to do it at first.

@DamianX the "gameplay" reason is in the body of the PR : 
>With just the frequency, you can talk, but you cannot hear. Useful for reporting your position with, say, a SBR hidden in your backpack tuned to sec without your abductors noticing shit.

You can basically already do that trick with a SBR with microphone on and receiver off.

greenhairedassistant commented:
fuck you, fuck you, fuck YOU.

stop nerfing smart players who plan ahead, were not bay. next you going to take away my yellow glove? FUCK YOU, stupid security player. if you dont want to people listen in, maybe stop being SHITCURITY and arresting people for no reason fuck you im going to be so mad if this get merges, security is designed to lose like in star trek the red shirts die so do security in ss13 stop making them tactical and strong and robust and op thats not what space station 13 is go back to bay if you want shitcurity rp fuckfests with snowflake statics. ss13 is about robust assistants and greytide fun and building rage cages, stop making security the center of your attention they suck and are supposed to suck that is the point of the game since it was created by goon and you bay turdlords are ruining it by making security not weak and adding voxes and aliens and skkeletons.

Inorien commented:
Frequency required to hear+speak, so if you guess or find out the frequency you can shitpost there as normal AND hear
What if a thing in tcomms allowed you to randomise the sec frequency such that all sec encryption keys would automatically update the headsets they're in, but headsets without encryption keys would have to find out/guess the new frequency
The hinderance in ability to meta frequencies would be balanced by increased attacks on sec to steal their headsets

ShiftyRail commented:
@Inorien this would defeat the entire point of having the Librarian start with the Security frequency in his memory.

Inorien commented:
Just have him telepathic, his brain is naturally tuned to all human communication in a 2 mile radius when he meditates in his office

Durgen commented:
nerf to greyshit and indirect buff to librarian 👌

ShiftyRail commented:
This last commit makes the use an encrypted key required to decode the channel. Without it, you can talk, without hearing.

I also refactored encryption so that pox raiders and nukies aren't snowflake checks anymore.

This creates a hefty amount of mapping issues, notably, the DJ sat, the listening outpost away mission, and of course the librarian who'll need to spawn with these keys to do his ace reporter job.
I've added the possibility to add an encryption key to intercoms and I'll see that the intercoms have one when it's needed.

From a balance standpoint, now that those keys have an actual use, maybe we should add a method to build them. I'm open to any (reasonable) idea. I would make it require NT Tech.

DamianX commented:
I still don't think you should be able to talk at all into a channel if you don't have the encryption key.

ShiftyRail commented:
You send your data, it's crypted, and you can only retrieve it if you can the decryption key when you're done.
I kinda see it as Public/Private key sort of thing, with the public key being in tcomms itself

ajc1g15 commented:
So if you tune to an encrypted channel, do you just hear nothing, or is it a garbled mess whenever someone speaks on it?

ShiftyRail commented:
@ajc1g15 Right now, you hear nothing. Same as if you were to listen to the raider/syndicate channel. I can make it so you hear a mess for flavour/immersion.

SonixApache commented:
hearing random static over the syndie channel during activity would be metagay as fuck

ajc1g15 commented:
Wouldn't it make more sense for the HoP to be able to do so? Given that he can assign people to jobs that would require stuff like that?

ShiftyRail commented:
@SonixApache do not forget that their frequency would be randomised.
If need be, we can make the raider and syndie channel completely blank.

@ajc1g15 for the HoP to do what ?

Accipitermaris commented:
I think designing Comms encryption such that messages sent over :S & :C _are_ encrypted when sent from a high security location but not when sent from anywhere else similar to our telescience teleprotection mechanic would be the most fun. This would still allow Greyshirts to listen in to cops on patrol which is a feature I like a lot while also allowing said cops to move into a secure location if they want to broadcast extremely sensitive data.

but heavens knows im not going to code that

nuklearcellphoneg commented:
>Probably not, but the Librarian will have it in his mind.

Ahahahaha

;Test wire is blue

;Librarian here, radio frequencies are...

SonixApache commented:
and thus the librarian was brigged for grand sabotage

Dacendeth commented:
Will I be able to stick an encryption key into a station intercom I built to listen to security?

ShiftyRail commented:
@nuklearcellphoneg it's at least a risk. Also, now we are using encryptions keys.
@Dacendeth yes. You can even insert it in your headset, it has 2 encryptions key slots.

DamianX commented on code/game/objects/items/devices/radio/intercom.dm:
```diff
@@ -53,25 +84,33 @@
 	spawn (0)
 		attack_self(user)
 
+#define CANT_RECIEVE
```
Should define this somewhere in __DEFINES, so that callers of this proc can use it as well.

DamianX commented on code/game/objects/items/devices/radio/radio.dm:
```diff
@@ -461,46 +459,52 @@
 	return
 */
 
+#define CANT_RECIEVE -1 // -1 = can't recieve for the caller of that proc
```
Should define this somewhere in __DEFINES, so that callers of this proc can use it as well.

DamianX commented on code/game/objects/items/devices/radio/encryptionkey.dm:
```diff
@@ -5,9 +5,9 @@
 	icon_state = "cypherkey"
 	item_state = ""
 	w_class = W_CLASS_TINY
+	origin_tech = Tc_MAGNETS + "=4;" + Tc_ENGINEERING + "=4;" + Tc_MATERIALS + "=4;" + Tc_PROGRAMMING + "=4"
```
Is this the new RnD speedrun?

ShiftyRail replied:
I honestly didn't have that in mind ! Thanks for the insight.

ShiftyRail commented:
This last commit adds encryptions keys in tcomms storage, namely, one security encryption key and one command encryption key. They are in plain slight, on the table.
Please react with :+1: if you think it's the right place, :confused: if you think they should be stored elsewhere, and :-1: if you believe those keys shouldn't be stored on station only be available when using a screwdriver on the headsets on-station.

ShiftyRail commented:
Whoops I forgot to add a paper to tcomms saying which freqs are what

SonixApache commented:
god bless the hunters of valid

aceedex commented:
Maybe a way to have pais use encryption keys too?

DamianX commented on vgstation13.dme:
```diff
@@ -2317,6 +2317,7 @@
 #include "interface\web\interface.dms"
 #include "maprendering\maprendering.dm"
 #include "maps\_map.dm"
+#include "maps\bagelstation.dm"
```
🤔

DamianX replied:
fix this
